habitat-lab/habitat/tasks/rearrange/utils.py

```diff
@@ -785,3 +606,197 @@ def get_camera_lookat_relative_to_vertial_line(
+    :property sim: The RearrangeSimulator or Simulator instance. This choice will dictate the collision detection routine.
+    :property pathfinder: The PathFinder defining the NavMesh to use.
+    :property target_object_id: An optional object_id which should be ignored in the occlusion check. For example, when pos is an object's COM, that object should not occlude the point.
+    :property island_id: Optionally restrict the search to a single navmesh island. Default -1 is the full navmesh.
+    :property search_offset: The additional radius to search for navmesh points around the target position. Added to the minimum distance from pos to navmesh.
+    :property test_batch_size: The number of sample navmesh points to consider when testing for occlusion.
+    :property max_samples: The maximum number of attempts to sample navmesh points for the test batch.
+    :property min_sample_dist: The minimum allowed L2 distance between samples in the test batch.
+    :property embodiement_heuristic_offsets: A set of 2D offsets describing navmesh cylinder center points forming a proxy for agent embodiement. Assumes x-forward, y to the side and 3D height fixed to navmesh. If provided, this proxy embodiement will be used for collision checking.
+    :property agent_embodiement: The MobileManipulator to be used for collision checking if provided.
+    :property orientation_noise: Standard deviation of the gaussian used to sample orientation noise. If 0, states always face the target point. Noise is applied delta to this "target facing" orientation.
+    :property max_orientation_samples: The number of orientation noise samples to try for each candidate point.
+    :property data_out: Optionally provide a dictionary which can be filled with arbitrary detail data for external debugging and visualization.
+
+    NOTE: this function is based on sampling and does not guarantee the closest point.
+
+    :return: A Tuple containing: 1) An approximation of the closest unoccluded snap point to pos or None if an unoccluded point could not be found, 2) the sampled orientation if found or None, 3) a boolean success flag.
+    """
+
+    assert height > 0
+    assert search_offset > 0
+    assert test_batch_size > 0
+    assert max_samples > 0
+    assert orientation_noise >= 0
+
+    if pathfinder is None:
+        pathfinder = sim.pathfinder
+
+    assert pathfinder.is_loaded
+
+    # first try the closest snap point
+    snap_point = pathfinder.snap_point(target_position, island_id)
+
+    # distance to closest snap point is the absolute minimum
+    min_radius = (snap_point - target_position).length()
+    # expand the search radius
+    search_radius = min_radius + search_offset
+
+    # gather a test batch
+    test_batch: List[Tuple[mn.Vector3, float]] = []
+    sample_count = 0
+    while len(test_batch) < test_batch_size and sample_count < max_samples:
+        sample = pathfinder.get_random_navigable_point_near(
```

Comment:
Do we know that this sample always returns a valid point? In other words, we do not want to have `NaN` cases